events/src/main/java/tech/pegasys/artemis/events/EventDeliverer.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.events;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import tech.pegasys.artemis.util.events.Subscribers;
+
+abstract class EventDeliverer<T> implements InvocationHandler {
+  private final Subscribers<T> subscribers = Subscribers.create(true);
+
+  void subscribe(T subscriber) {
+    subscribers.subscribe(subscriber);
+  }
+
+  @Override
+  public Object invoke(final Object proxy, final Method method, final Object[] args) {
+    if (method.getDeclaringClass().equals(Object.class)) {
+      try {
+        return method.invoke(this, args);
```

Comment:
Why do you do this call? It seems a bit unnecessary, to make sure the number of arguments is correct?

Reply:
So there are two sources of methods for an interface - those declared by the interface itself and those inherited from `Object`.  We only want to post events for the methods declared by the interface itself because the methods from `Object` require return values which we don't handle.

So for any method that's declared in the `Object` class we invoke the method on `this` which is the generated proxy object. Thus it winds up executing the default implementation fo those methods from `Object`.

Reply:
Gotcha, makes sense.